sign/ed25519/signapi.go

```diff
@@ -0,0 +1,84 @@
+package ed25519
+
+import (
+	"crypto/rand"
+	"encoding/asn1"
+
+	"github.com/cloudflare/circl/sign"
+)
+
+var Scheme sign.Scheme = &scheme{}
```

Comment:
I don't understand why to take the address of a zero-sized value. You can implement the interface with `scheme` and not with `*scheme`.

Reply:
Under the hood an instance of an interface is always a pair of a pointer and a concrete type, so instantiating it with a pointer is closer to what really happens.  The assembly generated btw is exactly [the same](https://godbolt.org/z/6zq31x).

Reply:
The difference is on the assembly calls.
     `call  AX` vs `call    "".(*Impl2).A(SB)`
one call is with a register and the other is with a pointer. 

I am not convinced why to use pointer when they are not needed.

Reply:
Oops, I simplified my example too much.

[Take a look at this one](https://godbolt.org/z/KsaccE)

Creating an instance of an interface with either a empty struct or a pointer to an empty struct results in the pointer `runtime.zerobase`.  (The code of `alloc1` and `alloc2` is the same.  This time I checked properly.)

The generic way to run a function on an interface involves a virtual call (`call ax`).  So interestingly the Go compiler optimised the call with the pointer receiver, but not the call with the value receiver.  I don't see a reason why it couldn't optimise the one with the value receiver as well and perhaps it will in the future.

So using pointers or not does not actually change anything when using empty structs (and the calling code doesn't know the types.) And if it does it optimizes the pointer case better. That's not the reason I use pointers though — as I said I prefer putting them here as it's closer to what Go actually does under the hood: an instance of an interface is always a pointer.

Reply:
> So interestingly the Go compiler optimised the call with the pointer receiver, but not the call with the value receiver.

I want to learn more about this.

> So using pointers or not does not actually change anything when using empty structs.

Agree.